services/brig/src/Brig/Federation/Client.hs

```diff
@@ -0,0 +1,76 @@
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Brig.Federation.Client where
+
+import Brig.API.Error (notFound, throwStd)
+import Brig.API.Handler (Handler)
+import Brig.App (federator)
+import Brig.Types.User
+import Control.Lens (view)
+import qualified Data.Aeson as Aeson
+import Data.Handle
+import Data.Qualified
+import Data.String.Conversions
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as LT
+import Imports
+import Mu.GRpc.Client.TyApps
+import qualified System.Logger.Class as Log
+import Wire.API.Federation.API.Brig
+import qualified Wire.API.Federation.GRPC.Types as Proto
+
+-- FUTUREWORK(federation): As of now, any failure in making a remote call results in 404.
+-- This is not correct, we should figure out how we communicate failure
+-- scenarios to the clients.
+-- See https://wearezeta.atlassian.net/browse/SQCORE-491 for the issue on error handling improvements.
+getUserHandleInfo :: Qualified Handle -> Handler (Maybe UserHandleInfo)
+getUserHandleInfo (Qualified handle domain) = do
+  Log.info $ Log.msg $ T.pack "Brig-federation: handle lookup call on remote backend"
+  federatorClient <- viewFederatorClient
+  let call = Proto.ValidatedRemoteCall domain (mkGetUserInfoByHandle handle)
+  res <- expectOk =<< callRemote federatorClient call
+  case Proto.responseStatus res of
+    404 -> pure Nothing
+    200 -> case Aeson.eitherDecodeStrict (Proto.responseBody res) of
+      Left err -> throwStd $ notFound $ "Failed to parse response: " <> LT.pack err
+      Right x -> pure $ Just x
+    code -> throwStd $ notFound $ "Invalid response from remote: " <> LT.pack (show code)
+
+viewFederatorClient :: Handler GrpcClient
+viewFederatorClient = do
+  mClient <- view federator
+  case mClient of
+    -- FUTUREWORK(federation): what happens if federator is transiently unreachable
+    -- at the time the grpc client is first initialized? Can we recover?
+    Nothing -> throwStd $ notFound "no federator configured or federator unreachable"
+    Just ep -> pure ep
+
+callRemote :: MonadIO m => GrpcClient -> Proto.ValidatedRemoteCall -> m (GRpcReply Proto.Response)
+callRemote fedClient call = liftIO $ gRpcCall @'MsgProtoBuf @Proto.RouteToRemote @"RouteToRemote" @"call" fedClient (Proto.validatedRemoteCallToRemoteCall call)
+
+expectOk :: GRpcReply Proto.Response -> Handler Proto.HTTPResponse
+expectOk = \case
+  GRpcTooMuchConcurrency _tmc -> throwStd $ notFound "Too much concurrency"
+  GRpcErrorCode errCode -> throwStd $ notFound $ LT.pack $ "GRPCError: " <> show errCode
+  GRpcErrorString errStr -> throwStd $ notFound $ LT.pack $ "GRPCError: " <> show errStr
+  GRpcClientError clErr -> throwStd $ notFound $ LT.pack $ "GRPC ClietnError: " <> show clErr
+  GRpcOk (Proto.ResponseErr err) -> throwStd $ notFound $ LT.pack $ "Remote component errored: " <> show err
+  GRpcOk (Proto.ResponseHTTPResponse res) -> pure res
```

Comment:
All of this code is only exercised in the test which goes between two Backends. This is not ideal, we should figure out a way to test client side of federated code without needing another backend. We could do this either by mocking the second backend in integration tests or making all of independent of the `Handler` monad and write unit tests.

Reply:
Made this comment into a futurework in https://github.com/wireapp/wire-server/pull/1319/commits/311cc14b7000bba4c28beced6e8868e92f984815